frontend/src/metabase/admin/permissions/components/tree/TreeNode.jsx

```diff
@@ -0,0 +1,92 @@
+import React from "react";
+import PropTypes from "prop-types";
+import {
+  TreeNodeRoot,
+  ExpandToggleButton,
+  ExpandToggleIcon,
+  NameContainer,
+  IconContainer,
+  RightArrowContainer,
+} from "./TreeNode.styled";
+
+import Icon from "metabase/components/Icon";
+
+const propTypes = {
+  isExpanded: PropTypes.bool.isRequired,
+  isSelected: PropTypes.bool.isRequired,
+  hasChildren: PropTypes.bool.isRequired,
+  onToggleExpand: PropTypes.func,
+  onSelect: PropTypes.func.isRequired,
+  depth: PropTypes.number.isRequired,
+  item: PropTypes.shape({
+    name: PropTypes.string.isRequired,
+    icon: PropTypes.string,
+    hasRightArrow: PropTypes.string,
+    id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
+  }).isRequired,
+};
+
+export const TreeNode = React.memo(function TreeNode({
+  isExpanded,
+  isSelected,
+  hasChildren,
+  onToggleExpand,
+  onSelect,
+  depth,
+  item,
+}) {
+  const { name, icon, hasRightArrow, id } = item;
+
+  const handleExpand = e => {
+    e.stopPropagation();
+    onToggleExpand(id);
+  };
+
+  const handleSelect = () => onSelect(id);
+
+  const handleKeyDown = ({ key }) => {
+    switch (key) {
+      case "Enter":
+        onSelect(id);
+        break;
+      case "ArrowRight":
+        !isExpanded && onToggleExpand(id);
+        break;
+      case "ArrowLeft":
+        isExpanded && onToggleExpand(id);
+        break;
```

Comment:
nice a11y -- I wonder how difficult it'd be to also implement an `ArrowUp` and `ArrowDown` action. I was playing with the page using my keyboard and it felt like it'd be a nice thing to have. No need to do it in this PR, though.

Reply:
Totally agree, I'll work on it separately